distributed_shampoo/distributed_shampoo.py

```diff
@@ -1142,6 +1148,114 @@ def _per_group_step_impl(
+            #   Example for AdaGrad accumulation:
+            #   L <- L + G * G^T
+            #   R <- R + G^T * G
+            #   V <- V + G^2    (element-wise)
+            #   (and similar)
+            self._update_preconditioners(
+                state_lists,
+                step,
+                grafting_config_not_none,
+            )
+
+            # Compute matrix root inverse.
+            #   L_inv <- L ** (-1/4)
+            #   R_inv <- R ** (-1/4)
+            #   (and similar)
+            self._compute_root_inverse(state_lists, compute_root_inverse)
+
+            # Compute filtered gradient or EMA of the gradients if beta1 > 0 and beta3 > 0.
+            # Note that we use two beta factors here akin to Lion.
+            #   G_bar <- beta3 * G_tilde + (1 - beta3) * G
+            #   G_tilde <- beta1 * G_tilde + (1 - beta1) * G
+            masked_filtered_grad_list = self._compute_filtered_grad_list(
+                state_lists,
+                step,
+                beta1,
+                beta3,
+                use_bias_correction,
+            )
+
+            # Precondition and graft filtered gradients.
+            # PT2 compile is currently disabled for preconditioning and grafting.
+            # TODO: Resolve preconditioning and grafting PT2 NEX issue and enable them.
+            #
+            #   P_shampoo <- L_inv * G_bar * R_inv (and similar)
+            #   P_grafting <- G_bar / (sqrt(V) + epsilon)
+            #   P <- P_grafting                                     if step < start_preconditioning_step
+            #   P <- ||P_grafting|| / ||P_shampoo|| * P_shampoo     otherwise
+            masked_blocked_search_directions = self._precondition_and_grafting(
+                state_lists,
+                masked_filtered_grad_list,
+                use_grafting_method,
+                grafting_config_not_none,
+            )
+
+        # Incorporate decoupled weight decay into search direction if enabled.
+        #   P <- P + weight_decay * W
+        self._apply_decoupled_weight_decay(
+            state_lists,
+            masked_blocked_search_directions,
+            weight_decay,
+            use_decoupled_weight_decay,
+        )
+
+        # Update momentum optimizer state and use momentum / Nesterov if enabled.
+        #   M <- momentum_param * M + (1 - dampening) * P
+        #   P <- (1 - dampening) * P + momentum_param * M     if use_nesterov
+        #   P <- M                                            otherwise.
+        self._update_momentum(
+            state_lists,
+            masked_blocked_search_directions,
+            momentum_param,
+            dampening,
+            use_nesterov,
+        )
+
+        # Updates parameters in distributed fashion.
+        # If DDP, executes AllGather communication to ensure all parameters are updated after local updates.
+        torch._foreach_mul_(masked_blocked_search_directions, neg_lrs)
```

Comment:
Note that this is exactly the same as the code for `_per_group_step_impl` with only the learning rate change, but this experimental path is just requiring us to pass in a list of tensors.